src/codegen/optimize.ml

```diff
@@ -134,6 +134,314 @@ let applicable_pattern p vars =
+      LetVal (v',(PVar fvar',ty',cont))
+    | (BigLambdaSkel (skparam,v), TySchemeSkel (skparam',ty), ApplySkelExp (_var,sk)) ->
+      assert (skparam = skparam');
+      Print.debug "BigLambdaSkel: replacing %t by %t" (CoreTypes.SkelParam.print skparam) (Types.print_skeleton sk);
+      let sub = Substitution.add_skel_param_substitution_e skparam sk in
+      (* v' = [t/s]v *)
+      let v' = Substitution.apply_substitutions_to_expression sub v in
+      (* T' = [t/s]T *)
+      let ty' = Substitution.apply_substitutions_to_type sub ty in
+      LetVal (v',(PVar fvar',ty',cont))
+    | (LambdaTyCoerVar (coparam,_ct_ty,v), QualTy (_ct_ty',ty), ApplyTyCoercion (_var,tyco)) ->
+      assert (_ct_ty = _ct_ty');
+      Print.debug "LambdaTyCoerVar: replacing %t by %t" (CoreTypes.TyCoercionParam.print coparam) (Typed.print_ty_coercion tyco);
+      let sub = Substitution.add_type_coercion_e coparam tyco in
+      (* v' = [y/w]v *)
+      let v' = Substitution.apply_substitutions_to_expression sub v in
+      (* T' = [y/w]T *)
+      let ty' = Substitution.apply_substitutions_to_type sub ty in
+      LetVal (v',(PVar fvar',ty',cont))
+    | (LambdaDirtCoerVar (coparam,_ct_dirt,v), QualDirt (_ct_dirt',ty), ApplyDirtCoercion (_var,dco)) ->
+      assert (_ct_dirt = _ct_dirt');
+      Print.debug "LambdaDirtCoerVar: replacing %t by %t" (CoreTypes.DirtCoercionParam.print coparam) (Typed.print_dirt_coercion dco);
+      let sub = Substitution.add_dirt_var_coercion_e coparam dco in
+      (* v' = [y/w]v *)
+      let v' = Substitution.apply_substitutions_to_expression sub v in
+      (* T' = [y/w]T *)
+      let ty' = Substitution.apply_substitutions_to_type sub ty in
+      LetVal (v',(PVar fvar',ty',cont))
+    | _ -> 
+      (* This should not happen, it should be handled by the match in `specialize_letrec`. *)
+      failwith __LOC__
+  in
+  (* Gather and substitute specializations *)
+  Print.debug "SPEC: specializing a letrec";
+  let cont' = specialize_comp cont in
+  (* Prefix the new specializations as let-bindings for cont. *)
+  Assoc.fold_left subst_spec cont' !assoc
+
+(* Drop unused bindings *)
+let letrec_drop_unused_bindings (c:computation) : computation =
+  match c with
+  | LetRec (bindings,c1) -> (
+      (* Get the free variables (function defs are the first part of the tuple)*)
+      let (free_vars_c1,_) = Typed.free_vars_comp c1 in
+      let free_vars_bindings = List.map (fun (_,_,_,a) -> fst (free_vars_abs a)) bindings in
+      let free_vars = List.flatten (free_vars_c1 :: free_vars_bindings) in
+      let used_bindings = List.filter (fun (var,_,_,_) -> List.mem var free_vars) bindings
+      in
+      (* if no bindings used, leave this LetRec out *)
+      match used_bindings with
+      | [] ->
+        Print.debug "Dropping letrec bindings";
+        c1
+      | _  -> LetRec (used_bindings, c1)
+    )
+  | _ -> (
+      Print.debug "This function can only be applied on LetRecs";
```

Comment:
It's best to pass the arguments of a LetRec separately then (would avoid the catch-all and the error)